packages/gatsby/src/internal-plugins/query-runner/pages-writer.js

```diff
@@ -35,13 +35,13 @@ const writePages = async () => {
     })
   })
 
-  pagesData = _.sortBy(
-    pagesData,
+  pagesData = _(pagesData)
+    // Ensure pages keep the same sorting through builds
+    .sortBy([`path`])
     // Sort pages with matchPath to end so explicit routes
     // will match before general.
-    p => (p.matchPath ? 1 : 0)
-  )
-
+    .sortBy(p => (p.matchPath ? 1 : 0))
```

Comment:
Could we combine those sorts, so we don't do additional work? test for `matchPath` should have higher priority than `path` sort.

Reply:
Yes I think it should be combined. I thought this way more understandable but sure less efficient. I change this.

Reply:
So how do think the following object should be sorted?

```js
[{
    path: '/lorem',
    matchPath: null,
}, {
    path: '/ipsum',
    matchPath: null,
}, {
    path: '/dolor',
    matchPath: '/foo',
}, {
    path: '/sit',
    matchPath: '/bar',
}, {
    path: '/amet',
    matchPath: null,
}]
```

Reply:
In general we want to create 2 groups - 1st that doesn't have `matchPath` (it would go in first) and 2nd that have `matchPath` and inside those group sort by `path` (ACS/DESC order doesn't matter here - what's matter that it's consistent).

So this would be
```
[{
    path: '/amet',
    matchPath: null,
}, {
    path: '/lorem',
    matchPath: null,
}, {
    path: '/ipsum',
    matchPath: null,
}, {
    path: '/dolor',
    matchPath: '/foo',
}, {
    path: '/sit',
    matchPath: '/bar',
}]
```

Reason we want pages with `matchPath` at the end is that this is used for dynamic paths and can be something like `/app/*` which would catch all `/app` routes. Reason we want those at the end is that if someone create another page with static path (say `/app/this-is-static`) we might never find it because it would be caught by dynamic page.

Reply:
Thanks for your accurate explanation. So I guess the current code works well. Then I will combine the two `sort` functions by paying attention to keep the same behavior.

Reply:
You can also adjust test fixture to make sure it works correctly ;)

Reply:
Would single sortBy like that work?
`.sortBy(p => (`${p.matchPath ? 1 : 0}${p.path}))`

Reply:
Thanks, it works fine.